File: src/main/java/ley/anvil/modpacktools/util/CustomToml.java
package ley.anvil.modpacktools.util;

import com.moandjiezana.toml.Toml;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.LinkedList;

public class CustomToml extends Toml {
    @SuppressWarnings("unchecked")
    public <T> T getPath(Class<T> type, String... path) {
        LinkedList<String> paths = new LinkedList<>(Arrays.asList(path));
        String key = paths.get(paths.size() - 1);
        paths.remove(paths.size() - 1);
        Toml toml = this;
        for(String str : paths)
            toml = toml.getTable(str);
        return (T)get(toml, key);
    }

    public Object get(Toml clazz, String key) {
        try {
            //Getting Around things being private for no reason 101
            Method method = Toml.class.getDeclaredMethod("get", String.class);
            method.setAccessible(true);
            return method.invoke(clazz, key);
        }catch(IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
            System.out.println("get fail");
            e.printStackTrace();
        }
        return null;
    }
}

File: src/main/kotlin/ley/anvil/modpacktools/util/CustomToml.kt
package ley.anvil.modpacktools.util

import com.moandjiezana.toml.Toml

class CustomToml : Toml() {
    companion object {
        @JvmStatic
        fun Toml.get(key: String): Any? {
            //Getting Around things being private for no reason 101 (dont look :P)
            val getFunc by lazy {Toml::class.getFun("get")}
            return getFunc?.call(this, key)
        }
    }

    /**
     * gets a path from a config.
     * when getting an Int do NOT supply int to T. instead supply Long and then convert to Int!
     *
     * @param T what to get from the config
     * @param path the path to get from the config separated by /, . or \
     */
    fun <T> getPath(path: String): T? = getPath(*path.split('/', '.', '\\').toTypedArray())

    /**
     * gets a path from a config.
     * when getting an Int do NOT supply int to T. instead supply Long and then convert to Int!
     *
     * @param T what to get from the config
     * @param path the path to get from the config
     */
    @Suppress("UNCHECKED_CAST")
    fun <T> getPath(vararg path: String): T? {
        var toml: Toml = this
        path.slice(0..path.size - 2).forEach {toml = toml.getTable(it)}
        return toml.get(path[path.size - 1]) as? T
    }
}